### PR TITLE
Add a metaclass implementation of __props__ and generalize it to all theano objects

### PR DESCRIPTION
**Pull request by abergeron**
This will remove problems with inheriting __eq__ or __hash__ that shadows our own.

### CODE REVIEW

File: theano/gof/utils.py [hunk 1]
```diff
@@ -152,22 +152,73 @@ class MethodNotDefined(Exception):
...
-            if hasattr(self, '__eq__') or hasattr(self, '__cmp__'):
-                raise TypeError("unhashable object: %s" % self)
-            return id(self)
 
     def __ne__(self, other):
         return not self == other
 
 
-class scratchpad:
+class scratchpad(object):
```
Comment (nouiz) [inline theano/gof/utils.py]: why this change? I suppose it is harmless, but I'm not 100% sure.

Comment (nouiz) [inline theano/gof/utils.py]: IRL @abergeron told that this is to be more consistent between python 2 and 3. In Python 3, it is like that by default.

Comment (abergeron) [inline theano/gof/utils.py]: This is to harmonize python 2 (where it switches between old-style and new-style) and python 3 (where it is a no-op).

**Comment by nouiz** [comment]
ok when both CI finish.

**Comment by Sentient07** [comment]
Hi @abergeron . From what I understand, we need to have `__props__` defined for all the OPs compulsorily? The Elemwise Op that takes a `dict` parameter, which is not hashable ( IIRC, `openmp`). What happens to those instances ?

**Comment by nouiz** [comment]
If an op don't have props, this metaclass does nothing.

Elemwise, override the \__hash__ and \__eq__ method. This is why it work
for them, it don't use the default version.
